Add tickless to some mbed-os devices

c1728p9 opened the pull request:
Add support for tickless by replacing RTX's SysTick timer code with with code which uses an mbed timer along with suspending and resuming the kernel in the idle loop.

c1728p9 commented:
/morph test

c1728p9 commented:
This is roughly based on #4956 but with SysTick replaced entirely.

c1728p9 commented:
/morph test

c1728p9 commented:
/morph test

bulislaw commented:
CC: @pan-

c1728p9 commented:
/morph test

0xc0170 commented on rtos/rtx5/mbed_rtx_idle.cpp:
```diff
@@ -0,0 +1,198 @@
+using namespace mbed;
+
+#if (defined(NO_SYSTICK))
+extern "C" IRQn_Type mbed_get_m0_tick_irqn(void);
```
Is this public API? targets that do not have systick (very few) should port this, should it be included in ticker api (is it related to tickers or own API just for nosystick rtos?)

missing the documentation what it should return

0xc0170 commented on rtos/rtx5/mbed_rtx_idle.cpp:
```diff
@@ -0,0 +1,198 @@
+    /**
+     * Get the time
+     *
+     * @return Current tmie in microseconds
```
typo `time`

c1728p9 replied:
done

0xc0170 commented on rtos/rtx5/mbed_rtx_idle.cpp:
```diff
@@ -0,0 +1,198 @@
+        schedule_tick();
+    }
+
+    us_timestamp_t start_time;
```
private members should be prefixed `_start_time`

c1728p9 replied:
done

0xc0170 commented on rtos/rtx5/mbed_rtx_idle.cpp:
```diff
@@ -0,0 +1,198 @@
+        // Ensure SysTick has the correct priority as it is still used
+        // to trigger software interrupts on each tick. The period does
+        // not matter since it will never start counting.
+        SysTick_Setup(16);
```
What does this do? I would expect to only set a new ticker vector for systick handler, nothing else, so why do we need this systick setup?

Looking at the weak function provided by RTX, its only used in osRtxSysTimerSetup.

c1728p9 replied:
This is only used to set the SysTick priority to the low value RTX expects (0xFF). I could put the code which sets the priority here, but to help with readability I called SysTick_Setup instead. Setting the priority is slightly different for each cortex variation and would look something like this:


```
#if   ((__ARM_ARCH_8M_MAIN__ == 1U) || (defined(__CORTEX_M) && (__CORTEX_M == 7U)))
  SCB->SHPR[11] = 0xFFU;
#elif  (__ARM_ARCH_8M_BASE__ == 1U)
  SCB->SHPR[1] |= 0xFF000000U;
#elif ((__ARM_ARCH_7M__      == 1U) || \
       (__ARM_ARCH_7EM__     == 1U))
  SCB->SHP[11]  = 0xFFU;
#elif  (__ARM_ARCH_6M__      == 1U)
  SCB->SHP[1]  |= 0xFF000000U;
#endif
```

0xc0170 commented on rtos/rtx5/mbed_rtx_idle.cpp:
```diff
@@ -0,0 +1,198 @@
+            // Don't update to the current tick.
+            // The RTOS must do this to properly
+            // schedule tasks again.
+            new_tick--;
```
It's not clear to me why we do `-1` in case `newtick > tick` ? If ticks have increased since the last update tick, we substract one from the tick count?

c1728p9 replied:
The function svcRtxKernelResume doesn't trigger a task switch if a thread is ready to run. To work around this and jump start scheduling after a tick has passed, one less than the number of ticks that have occurred are passed to svcRtxKernelResume and SysTick is set to pending. The SysTick interrupt then increments the os's tick count to the right value and performs task switching as normal.

After further analysis, it appears that PendSV may be more appropriate in this situation. It will trigger a thread switch without recomputing the tick count. I'll update the PR to use this instead.

0xc0170 commented on rtos/rtx5/mbed_rtx_idle.cpp:
```diff
@@ -0,0 +1,198 @@
+static uint64_t os_timer_data[sizeof(RtosTimer) / 8];
+
+/// Setup System Timer.
+int32_t osRtxSysTimerSetup (void)
```
Isn't there CMSIS-RTOS for tick API? What I found is https://github.com/ARM-software/CMSIS_5/blob/6e5621b3bc4becf50ac5f29a07f30ff2081a5e70/CMSIS/RTOS2/Include/os_tick.h , but that seems to be not in our codebase - https://www.keil.com/pack/doc/CMSIS/RTOS2/html/group__CMSIS__RTOS__TickAPI.html. I guess we get this once we update RTOS

Just to be aware

bulislaw replied:
They've changed the names again, the API was there and is used by nrf51_dk https://github.com/ARMmbed/mbed-os/blob/5f138810a90e913f617f4345654709eb11281385/rtos/rtx5/TARGET_CORTEX_M/rtx_system.c#L212

bulislaw commented:
Fails for Ameba:
```
[1504175141.97][CONN][RXD] >>> Running case #7: 'Testing single thread with wait'...
[1504175141.98][CONN][INF] found KV pair in stream: {{__testcase_start;Testing single thread with wait}}, queued...
[1504175142.02][CONN][RXD] mbed assertation failed: os_timer->get_tick() == svcRtxKernelGetTickCount(), file: /Users/barsza01/devel/mbed/code/mbed-os/rtos?V?V??(Q???V??(Q
```

bulislaw commented:
@Archcady this PR fails for Ameba board, I've tried to see what the platform does around the SysTick, but I don't think it's int the sources (unless i've missed it). Can you have a look and see why would it fail?

c1728p9 commented:
> If the SysTick is tied to the us_ticker then an additional latency might be present. Is it a good thing ?

This will cause the additional latency of a second interrupt call, but this should be fairly small when compared to a 1ms tick. The reason this is done is that the interrupt that runs SysTick must be run at a low priority. On most if not all platforms, the lp/us ticker interrupt is set to a higher priority than this.

> RtosTimer shall take into account OS_TICK_FREQ.

Done

> It would be good to have a way to override the behavior in targets, the implementation exposed, on 
NRF51, is I believe significantly less precise than the old one (it would be good to benchmark it to verify that assertion).

I'll compare this to the time reported by the us ticker as we discussed. Unfortunately I don't have a board with me so this might have to wait until another day.

c1728p9 commented:
/morph test

c1728p9 commented:
Summary of updates:
* Addressed PR feedback
* Switched from to the US to LP ticker
* Removed tick -1 code and now use SetPendSV instead
* Put tickless behind MBED_TICKLESS macro
* Removed tick comparison assert (Disabling interrupts for too long (2ms+) in application can cause a SysTick interrupt to get dropped which triggers this assert)

bulislaw commented:
Both Ameba and GG boards are fixed. Great job @c1728p9 :)

bulislaw commented:
Ah that's because it's now disabled by default :)

0xc0170 commented:
@pan- Can you review again please?

c1728p9 commented:
Turned off tickless on the F401 since this device does not have a lp ticker.

c1728p9 commented:
/morph test

c1728p9 commented:
Fixed a regression by moving the PendSV call outside the critical section. Also rebased to master.

c1728p9 commented:
/morph test

c1728p9 commented:
@pan- I profiled this on the nrf51 whe the below program and I didn't see any drift.
```c
#include "mbed.h"

int main()
{
    uint32_t next_wakeup = 1000;
    while (1) {
        osDelayUntil(next_wakeup);
        printf("lp ticker: %lu\r\n", lp_ticker_read());
        next_wakeup += 1000;
    }
    return 0;
}

```

Console output:
```
lp ticker: 1000641
lp ticker: 2000641
lp ticker: 3000641
lp ticker: 4000641
lp ticker: 5000641
...
lp ticker: 160000641
lp ticker: 161000641
lp ticker: 162000641
lp ticker: 163000641
```

bulislaw commented:
/morph test

c1728p9 commented:
/morph test

c1728p9 commented:
-reverted PendSV change since this causes a scheduling lockup if the background threads is the only thread that is ready to run
-Reschedule the tick in osRtxSysTimerAckIRQ so ticks cannot be dropped
-Add tick assert back since ticks can never be dropped

c1728p9 commented:
* Removed tickless support for NUCLEO_F429ZI and MAX32625MBED
* Rebased to master
* Removed space savings for stm32f07
* moved mbed_rtx_idle.cpp behind TARGET_CORTEX since restructure PR was merged

c1728p9 commented:
/morph test

c1728p9 commented:
Rebased to master again and added back space savings for stm32f07. This is required to compile since the background thread has been bumped for all devices.

c1728p9 commented:
/morph test

bulislaw commented:
@studavekar I think somethings wrong with the CI pipeline. This PR is 'must have' for next release can we prioritise when the CI is unclogged?

theotherjimmy commented:
@c1728p9 NUCLEO_F07xRB failed by not fitting in their ROM.

theotherjimmy commented:
@bulislaw Are there any more "must haves"? If so, could you label them with 5.6.0? We can better prioritize them that way.

0xc0170 commented:
Wit this one, I asked in the second implementation ,which one should go in, we should not run 2x CI if 1 is not neccesary (at least at this stage).

bulislaw commented:
> @c1728p9 NUCLEO_F07xRB failed by not fitting in their ROM.

That's not true, that's the previous results. For 1237 that's the new build matrix, all seem to be success http://mbed-ci-master-2.austin.arm.com:8081/job/build_matrix/1102/ but the run seems to be stuck

theotherjimmy commented:
@bulislaw Yes, I was referring to run 1235.

c1728p9 commented:
/morph test

studavekar commented:
/morph test

c1728p9 commented:
rebased to master to fix conflicts

c1728p9 commented:
/morph test

bulislaw commented:
@pan- are you ok with the state of things?

adbridge commented:
Closes and re-opened to kick off travis again

saedelman commented on rtos/TARGET_CORTEX/mbed_rtx_idle.cpp:
```diff
@@ -0,0 +1,230 @@
+
+    core_util_critical_section_enter();
+    uint32_t ticks_to_sleep = svcRtxKernelSuspend();
+    MBED_ASSERT(os_timer->get_tick() == svcRtxKernelGetTickCount());
```
breaks debugger

0xc0170 replied:
@saedelman Can you provide more details, or create an issue to reference this line there so we can review?

saedelman replied:
Using gdb and single stepping through the RTX code causes the assert to be triggered. This may be related to the "Gotcha" related to in https://github.com/nuket/mbed-memory-status although I am using SEGGER RTT debug output not serial output:

"On mbed 5.5 and up, there may be a Heisenbug when calling print_thread_info() inside of osKernelLock()!

This error sometimes appears on the serial console shortly after chip startup, but not always: mbed assertation failed: os_timer->get_tick() == svcRtxKernelGetTickCount(), file: .\mbed-os\rtos\TARGET_CORTEX\mbed_rtx_idle.c

The RTOS seems to be asserting an idle constraint violation due to the slowness of sending data through the serial port, but it does not happen consistently."
